### PR TITLE
Checklists – Separate manual and generated risks in UI logic

### PR DESCRIPTION
**Pull request by lmcnulty**
Resolves https://github.com/responsible-ai-collaborative/aiid/issues/2650.

### CODE REVIEW

File: site/gatsby-site/src/components/checklists/CheckListForm.js
Comment (kepae) [inline site/gatsby-site/src/components/checklists/CheckListForm.js:166]: I'm wary of all the `|| []`. Nothing to do here right now but we are hiding empty values all over the place. Please consider this worth coming back to, after this re-factor.

**Review by kepae — COMMENTED** [review]
Overall, great going on the re-factor and making this easier to maintain. In next changes, we need to be more careful with the data structures flying around -- there's a lot of mental overhead here and we're hiding a lot of potential errors with `|| []`.

What would be the best way to test these changes? both in our test suite and in demo?
We can follow up with tests in a separate PR, but we should know this works before deploying.

File: site/gatsby-site/src/components/checklists/RiskSections.js
Comment (lmcnulty) [reply site/gatsby-site/src/components/checklists/RiskSections.js:31]: Yes, those are the generated risks that should not be displayed, because they are covered by the manual risks.

File: site/gatsby-site/src/components/checklists/RiskSections.js
Comment (lmcnulty) [reply site/gatsby-site/src/components/checklists/RiskSections.js:47]: Everything stored in `risks` is a manual risk. I could rename that to be clearer.

**Comment by lmcnulty** [comment]
Not, there are some obselete fields in the checklists realm schema, like `risks[].generated`. I'm not sure I want to remove them until after this is merged, because removing fields from schemas tends to make switching branches annoying if they're not synced in what they expect.

**Comment by kepae** [comment]
Agree, worry about removing fields after this is merged, preferring to do it asap though. I would add a todo in the code and consider it tombstoned.

Per our discussion yesterday, let me know when there are some new inline documentation and readability changes, especially regarding:
* what is the de-duplicating strategy
* how the strategy is implemented, e.g. how does "fingerprinting" risk objects to detect equivalence work, why is this sufficient

**Comment by lmcnulty** [comment]
There's some naming tension here.

Right now, the generated risks returned from the API are in a response field called `risks`. The manually-annotated risks actually stored in the database are also called `risks`. In the code right now, the `risks` form field stores the manually-annotated `risks`, whereas the generated risks are stored in a list called `generatedRisks`. This is all kind of confusing.

Here are some other potential names that come to mind for the manually-annotated risks stored in the database. Ideally the name should indicate that they are stored in the database at human discretion.

-   risks
-   manualRisks
-   riskAnnotations
-   riskEntries
-   annotatedRisks
-   knownRisks
-   manuallyAnnotatedRisks

And here are some that come to mind for the potential risks dynamically generated based on failure annotations matching the checklist query tags. Ideally I would like the name to indicate that these are automatically-generated, change over time, and are not necessarily practical risks of the system involved – in the end they're just query results of matching tags.

-   risks
-   generatedRisks
-   potentialRisks
-   precedentedRisks
-   surfacedRisks
-   riskPreviews
-   riskCandidates
-   candidateRisks
-   riskPotentials
-   surfacedFailures
-   knownFailures
-   relatedFailures
-   riskResults

The name for both should also contrast enough that it's clear that they're not the same data type. "generated vs manual" does not accomplish that.

I think right now I'm leaning toward "riskAnnotations" vs "surfacedRisks". The fact that "risk" is in both words tells us that they're related, but using it as an adjective in one and a noun in the other tells us that they're not the same data type: one is a risk, the other is an annotation about a risk.

I'd like to hear another opinion on this before I change the names everywhere.

File: site/gatsby-site/src/components/checklists/CheckListForm.js
Comment (lmcnulty) [reply site/gatsby-site/src/components/checklists/CheckListForm.js:166]: I had previously added a bunch of `|| []`s because we were getting crashes in unexpected, hard-to-diagnose circumstances. I would like to be able to guarantee that we'll never get a null where we're not expecting one, but I don't really know how to that in JavaScript. This seems like something there would be best practices about, but I don't really know what they are beyond the fact that things like this are considered suspect.

File: site/gatsby-site/src/utils/checklists.js
Comment (lmcnulty) [reply site/gatsby-site/src/utils/checklists.js:75]: In this case, it's just because it's clutter and I don't think most users would want it. In other places we were removing `__typename` because it used to be the case that including `__typename`s in an updated value would cause a schema validation error. That's no longer true, so I took it out everywhere but here.

File: site/gatsby-site/src/utils/checklists.js
Comment (lmcnulty) [reply site/gatsby-site/src/utils/checklists.js:63]: There's only three places where a risk can be created, and in each case some of the fields get overridden. I think it's clearest just to create the full object inline in each place.

**Comment by kepae** [comment]
> I think right now I'm leaning toward "riskAnnotations" vs "surfacedRisks". The fact that "risk" is in both words tells us that they're related, but using it as an adjective in one and a noun in the other tells us that they're not the same data type: one is a risk, the other is an annotation about a risk.

This makes perfect sense to me. We'll always have some amount of confusing naming when it comes to UIs handling data stateful data structures, but I think this is sufficiently readable. (I don't see a more readable option myself.)

Also, the new function comments make it easier to read this (even with some time elapsed!).

**Comment by kepae** [comment]
As discussed on Slack, let's put this into staging since it's supposed to help us address existing maintainability, and consider further renaming any variables later. The comments already help readability massively, given the logic.

**Review by kepae — APPROVED** [review]

**Pull request merged**
